### PR TITLE
DAOS-6246 hlc: Optimize timestamp format

### PR DESCRIPTION
**Pull request by liw**
Optimize the HLC timestamp format, reducing the overall representable
time range for a better resolution of the physical portion and a larger
range of the logical portion. See DAOS-6246 for more details.

  - Change the HLC timestamp unit from 1 ns to 1/16 ns, and change the
    time represented by HLC timestamp 0 from the Unix Epoch to
    "2021-01-01 00:00:00 +0000 UTC". This gives a range of [2021, 2057].

  - Change the number of logical bits from 16 to 18. In addition to more
    logical bits, this also improves the resolution of the physical
    portion from about 64 us to 16 us.

Signed-off-by: Li Wei <wei.g.li@intel.com>

### CODE REVIEW

**Comment by liw** [comment]
daos_core_test -T dtx_37: DAOS-6538

**Comment by liw** [comment]
daos_test -z timed out: DAOS-6507

File: src/cart/crt_hlc.c [hunk 1]
```diff
@@ -28,12 +28,25 @@
...
+ * HLC timestamp unit (given in the HLC timestamp value for 1 ns) (i.e.,
+ * 1/16 ns, offering a 36-year range)
+ */
+#define CRT_HLC_NSEC 16ULL
+
+/**
+ * HLC start time (given in the Unix time for 2021-01-01 00:00:00 +0000 UTC in
+ * seconds) (i.e., together with CRT_HLC_NSEC, offering a range of [2021, 2057])
+ */
+#define CRT_HLC_START 1609459200ULL
```
Comment (frostedcmos) [inline src/cart/crt_hlc.c:41]: maybe change name to CRT_HLC_START_SEC  for clarity

File: src/cart/crt_hlc.c [hunk 2]
```diff
@@ -96,9 +109,35 @@ int crt_hlc_get_msg(uint64_t msg, uint64_t *hlc_out, uint64_t *offset)
...
 }
 
-uint64_t crt_hlc2sec(uint64_t hlc)
+uint64_t crt_hlc2nsec(uint64_t hlc)
 {
-	return (hlc & ~CRT_HLC_MASK) / NSEC_PER_SEC;
+	return hlc / CRT_HLC_NSEC;
+}
+
+uint64_t crt_hlc_from_nsec(uint64_t nsec)
```
Comment (frostedcmos) [inline src/cart/crt_hlc.c:117]: why not crt_nsec2hlc ?

File: src/cart/crt_hlc.c [hunk 3]
```diff
@@ -96,9 +109,35 @@ int crt_hlc_get_msg(uint64_t msg, uint64_t *hlc_out, uint64_t *offset)
...
+{
+	return nsec * CRT_HLC_NSEC;
+}
+
+uint64_t crt_hlc2unixnsec(uint64_t hlc)
+{
+	return hlc / CRT_HLC_NSEC + CRT_HLC_START * NSEC_PER_SEC;
+}
+
+int crt_hlc_from_unixnsec(uint64_t unixnsec, uint64_t *hlc)
```
Comment (frostedcmos) [inline src/cart/crt_hlc.c:127]: same here why not crt_unixnsec2hlc for symmetry?

File: src/cart/crt_hlc.c [hunk 4]
```diff
@@ -96,9 +109,35 @@ int crt_hlc_get_msg(uint64_t msg, uint64_t *hlc_out, uint64_t *offset)
...
+{
+	uint64_t start = CRT_HLC_START * NSEC_PER_SEC;
+
+	/*
+	 * If the time represented by unixnsec is before the time represented by
+	 * CRT_HLC_START, then the conversion is impossible.
+	 */
+	if (unixnsec < start)
+		return -DER_INVAL;
```
Comment (frostedcmos) [inline src/cart/crt_hlc.c:136]: why not return 0 on error and actual value on success? would make usage easier, and return value of 0 is impossible for valid inputs

File: src/cart/crt_hlc.c (hunk 1 above)
Comment (liw) [reply src/cart/crt_hlc.c:41]: OK.

File: src/cart/crt_hlc.c (hunk 2 above)
Comment (liw) [reply src/cart/crt_hlc.c:117]: OK. (crt_nsec2hlc doesn't start with "crt_hlc_". I generally avoid "2" in function names, and find x_to_y and x_from_y favorable.)

File: src/cart/crt_hlc.c (hunk 3 above)
Comment (liw) [reply src/cart/crt_hlc.c:127]: OK.

File: src/cart/crt_hlc.c (hunk 4 above)
Comment (liw) [reply src/cart/crt_hlc.c:136]: If the input corresponds to CRT_HLC_START_SEC, then the output is a valid 0.

Comment (frostedcmos) [reply src/cart/crt_hlc.c:136]: sure, but cant we impose that input has to be above start_sec time?the real clocks should all be now above jan 1 2021 00:00:00

Comment (liw) [reply src/cart/crt_hlc.c]: OK, changed.

**Review by frostedcmos — DISMISSED** [review]

**Review by johannlombardi — APPROVED** [review]

**Comment by liw** [comment]
Merged master to resolve "copyright" conflicts.

**Review by frostedcmos — DISMISSED** [review]

**Review by johannlombardi — DISMISSED** [review]

**Comment by johannlombardi** [comment]
Is the testing stuck?

**Comment by frostedcmos** [comment]
> Is the testing stuck?

it probably was remerged at a bad time when unit tests were failing last week; other stages of pipeline wont execute if unit tests failed, and 5 did here. probably best to remerge with latest master again and retry

**Comment by liw** [comment]
osa_online_drain: DAOS-6664

**Comment by liw** [comment]
No code changes since last round of reviews.

**Review by frostedcmos — APPROVED** [review]

**Review by johannlombardi — APPROVED** [review]

**Pull request merged**